llvm/unittests/CAS/LazyAtomicPointerTest.cpp

```diff
@@ -34,26 +35,29 @@ TEST(LazyAtomicPointer, loadOrGenerate) {
   EXPECT_EQ(Ptr.load(), &Value);
 }
 
+#if (LLVM_ENABLE_THREADS)
 TEST(LazyAtomicPointer, BusyState) {
   int Value = 0;
   LazyAtomicPointer<int> Ptr;
   ThreadPool Threads;
 
-  std::mutex BusyStart, BusyEnd;
-  BusyStart.lock();
-  BusyEnd.lock();
+  std::mutex BusyLock, EndLock;
+  std::condition_variable Busy, End;
   Threads.async([&]() {
     Ptr.loadOrGenerate([&]() {
-      BusyStart.unlock();
-      while (!BusyEnd.try_lock()) {
-        // wait till the lock is unlocked.
-      }
+      // Notify busy state.
+      Busy.notify_all();
```

Comment:
I think you need to acquire `BusyLock` before notifying the condition.

Reply:
Well, I can't lock before this. That will be a deadlock. 

execute [notify_one](https://en.cppreference.com/w/cpp/thread/condition_variable/notify_one) or [notify_all](https://en.cppreference.com/w/cpp/thread/condition_variable/notify_all) on the std::condition_variable (the lock does not need to be held for notification)

That doesn't seem necessary?

Reply:
Here's an example of how this should look: